File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/21-HashTable/LeetCode/LeetCode-0706.java
class Pair {
    int key;
    int value;
    public Pair(int key, int value){
        this.key = key;
        this.value = value;
    }
}

class MyHashMap {
    private int TABLE_SIZE = 2011; // Some prime number
    private List<Pair>[] T = new LinkedList[TABLE_SIZE];

    public MyHashMap() {}
    
    public void put(int key, int value) {
        int index = key % TABLE_SIZE;
        if (T[index] == null) T[index] = new LinkedList<>();
        for (var pair: T[index]){
            if (pair.key == key){
                pair.value = value;
                return;
            }
        } // end-for

        T[index].add(new Pair(key, value));
    }
    
    public int get(int key) {
        int index = key % TABLE_SIZE;
        if (T[index] == null) return -1;

        for (var pair: T[index])
            if (pair.key == key)
                return pair.value;

        return -1;        
    }
    
    public void remove(int key) {
        int index = key % TABLE_SIZE;
        if (T[index] == null) return;
        for (int i=0; i<T[index].size(); i++){
            if (T[index].get(i).key == key){
                T[index].remove(i);
                return;
            } // end-for
        } // end-for
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/21-HashTable/LeetCode/LeetCode-0706.cpp
class MyHashMap {
    #define TABLE_SIZE 2011 // 1013 //5011    // Some prime number
    vector<list<pair<int, int>>> T;

public:
    /** Initialize your data structure here. */
    MyHashMap() {T.resize(TABLE_SIZE);}

    /** value will always be non-negative. */
    void put(int key, int value) {
        int index = key % TABLE_SIZE;
        for (auto iter = T[index].begin(); iter != T[index].end(); iter++){
            if (iter->first == key){
                iter->second = value;
                return;
            } //end-if
        } //end-for

        T[index].push_back(make_pair(key, value));
    } // end-put

    /** Returns the value to which the specified key is mapped,
        or -1 if this map contains no mapping for the key */
    int get(int key) {
        int index = key % TABLE_SIZE;
        for (auto p: T[index]){
            if (p.first == key) return p.second;   // Exists. Return the value
        } //end-for

        return -1;    // Does not exist
    } //end-get

    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    void remove(int key) {
        int index = key % TABLE_SIZE;
        for (auto iter = T[index].begin(); iter != T[index].end(); iter++){
            if (iter->first == key){
                T[index].erase(iter);
                return;
            } //end-if
        } //end-for
    } //end-remove
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */